test/e2e/addon_update.go

```diff
@@ -0,0 +1,266 @@
+		sshExecAndVerify(sshClient, fmt.Sprintf("sudo rm %s/%s", destinationDir, rcv1))
+		sshExecAndVerify(sshClient, fmt.Sprintf("sudo rm %s/%s", destinationDir, svcv1))
+		/**
+		 * Note that we have a small race condition here - the kube-addon-updater
+		 * May notice that a new rc/service file appeared, while the old one will still be there.
+		 * But it is ok - as long as we don't have rolling update, the result will be the same
+		 */
+
+		waitForService(c, "addon-test-updated", true)
+		waitForReplicationController(c, "addon-test-v2", true)
+
+		waitForService(c, "addon-test", false)
+		waitForReplicationController(c, "addon-test-v1", false)
+
+		By("remove manifests")
+		sshExecAndVerify(sshClient, fmt.Sprintf("sudo rm %s/%s", destinationDir, rcv2))
+		sshExecAndVerify(sshClient, fmt.Sprintf("sudo rm %s/%s", destinationDir, svcv2))
+
+		waitForService(c, "addon-test-updated", false)
+		waitForReplicationController(c, "invalid-addon-test-v1", false)
+
+		By("verify invalid API addons weren't created")
+		_, err = c.ReplicationControllers(addonNamespace).Get("invalid-addon-test-v1")
+		Expect(err).To(HaveOccurred())
+		_, err = c.Services(addonNamespace).Get("ivalid-addon-test")
+		Expect(err).To(HaveOccurred())
+
+		// invalid addons will be deleted by the deferred function
+	})
+})
+
+func waitForService(c *client.Client, name string, exist bool) {
+	err := wait.Poll(addonTestPollInterval, addonTestPollTimeout, func() (bool, error) {
+		_, err := c.Services(addonNamespace).Get(name)
+		if err != nil {
+			Logf("Get service %s in namespace %s failed (%v).", name, addonNamespace, err)
+			return !exist, nil
+		} else {
+			Logf("Service %s in namespace %s found.", name, addonNamespace)
+			return exist, nil
+		}
+	})
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func waitForReplicationController(c *client.Client, name string, exist bool) {
+	err := wait.Poll(addonTestPollInterval, addonTestPollTimeout, func() (bool, error) {
+		_, err := c.ReplicationControllers(addonNamespace).Get(name)
+		if err != nil {
+			Logf("Get ReplicationController %s in namespace %s failed (%v).", name, addonNamespace, err)
+			return !exist, nil
+		} else {
+			Logf("ReplicationController %s in namespace %s found.", name, addonNamespace)
+			return exist, nil
+		}
+	})
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func getSSHClient() (*ssh.Client, error) {
+	// Get a signer for the provider.
+	signer, err := getSigner(testContext.Provider)
+	if err != nil {
+		return nil, fmt.Errorf("error getting signer for provider %s: '%v'", testContext.Provider, err)
+	}
+
+	config := &ssh.ClientConfig{
+		User: os.Getenv("USER"),
+		Auth: []ssh.AuthMethod{ssh.PublicKeys(signer)},
+	}
+
+	host := getMasterHost() + ":22"
+	client, err := ssh.Dial("tcp", host, config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting SSH client to host %s: '%v'", host, err)
+	}
+	return client, err
+}
+
+func sshExecAndVerify(client *ssh.Client, cmd string) {
```

Comment:
This should be moved out to utility-land at some point - I assume there's other uses.